tracer/src/Datadog.Trace/Telemetry/Collectors/MetricsTelemetryCollector.cs

```diff
@@ -242,73 +262,113 @@ public void Clear()
         return metricKeyTags;
     }
 
-    private record struct MetricKey
+    private struct AggregatedMetric
     {
-        public int Value;
+        private List<MetricDataPoint>? _values;
         public readonly string[]? Tags;
 
-        public MetricKey(string[]? tags)
+        public AggregatedMetric(string[]? tags)
         {
-            Value = 0;
             Tags = tags;
         }
+
+        public bool HasValues => _values is { Count: > 0 };
+
+        public void AddValue(long timestamp, int value)
+        {
+            // we set the default capacity to 8 because we expect to get 6 values (steady state)
+            // but we may get a couple of extra values initially, and don't want to end up needing to expand
+            _values ??= new(capacity: 8);
+            _values.Add(new(timestamp, value));
+        }
+
+        public MetricSeries? GetAndClear()
+        {
+            if (_values is { Count: > 0 } values)
+            {
+                var result = new MetricSeries(values);
+                values.Clear();
+                return result;
+            }
+
+            return null;
+        }
     }
 
-    private record struct DistributionKey
+    private struct AggregatedDistribution
     {
-        public BoundedConcurrentQueue<double> Values;
+        private List<double>? _values;
         public readonly string[]? Tags;
 
-        public DistributionKey(string[]? tags)
+        public AggregatedDistribution(string[]? tags)
         {
-            Values = new(queueLimit: 1000);
             Tags = tags;
         }
+
+        public bool HasValues => _values is { Count: > 0 };
+
+        public void AddValue(double value)
+        {
+            _values ??= new();
```

Comment:
It's a bit dangerous with a struct. 

If a copy is made before adding the first value, both instances of AggregatedDistribution will have different points.
If a copy is made after adding the first value, both instances of AggregatedDistribution will have the same points.

It's probably OK-ish because the struct is private, but that's error prone.

Reply:
Discussed offline using a `readonly struct` (and `WithValue` helper etc) and other options, but the easiest "fix" is just to use a `class`. Even though we allocate a lot of them, it should have a relatively minor impact given these are only allocated once.